bson/types.go

```diff
@@ -7,44 +7,111 @@
 package bson
 
 import (
-	"go.mongodb.org/mongo-driver/bson/bsontype"
+	"encoding/json"
+	"net/url"
+	"reflect"
+	"time"
+
+	"go.mongodb.org/mongo-driver/x/bsonx/bsoncore"
 )
 
+// Type represents a BSON type.
+type Type bsoncore.Type
+
+// String returns the string representation of the BSON type's name.
+func (bt Type) String() string {
+	return bsoncore.Type(bt).String()
+}
+
+// IsValid will return true if the Type is valid.
+func (bt Type) IsValid() bool {
+	switch bt {
+	case TypeDouble, TypeString, TypeEmbeddedDocument, TypeArray, TypeBinary,
+		TypeUndefined, TypeObjectID, TypeBoolean, TypeDateTime, TypeNull, TypeRegex,
+		TypeDBPointer, TypeJavaScript, TypeSymbol, TypeCodeWithScope, TypeInt32,
+		TypeTimestamp, TypeInt64, TypeDecimal128, TypeMinKey, TypeMaxKey:
+		return true
+	default:
+		return false
+	}
+}
+
 // BSON element types as described in https://bsonspec.org/spec.html.
 const (
-	TypeDouble           = bsontype.Double
-	TypeString           = bsontype.String
-	TypeEmbeddedDocument = bsontype.EmbeddedDocument
-	TypeArray            = bsontype.Array
-	TypeBinary           = bsontype.Binary
-	TypeUndefined        = bsontype.Undefined
-	TypeObjectID         = bsontype.ObjectID
-	TypeBoolean          = bsontype.Boolean
-	TypeDateTime         = bsontype.DateTime
-	TypeNull             = bsontype.Null
-	TypeRegex            = bsontype.Regex
-	TypeDBPointer        = bsontype.DBPointer
-	TypeJavaScript       = bsontype.JavaScript
-	TypeSymbol           = bsontype.Symbol
-	TypeCodeWithScope    = bsontype.CodeWithScope
-	TypeInt32            = bsontype.Int32
-	TypeTimestamp        = bsontype.Timestamp
-	TypeInt64            = bsontype.Int64
-	TypeDecimal128       = bsontype.Decimal128
-	TypeMinKey           = bsontype.MinKey
-	TypeMaxKey           = bsontype.MaxKey
+	TypeDouble           = Type(bsoncore.TypeDouble)
+	TypeString           = Type(bsoncore.TypeString)
+	TypeEmbeddedDocument = Type(bsoncore.TypeEmbeddedDocument)
```

Comment:
These BSON type constants are based on the BSON spec, so it's safe to duplicate them in both the `bson` package and the `bsoncore` package.